stdlib/buffer.ml

```diff
@@ -73,7 +74,12 @@ let resize b more =
   Bytes.blit b.buffer 0 new_buffer 0 b.position;
```

Comment:
I don't understand why some of the occurrences of `b.position` are changed to `old_pos` and some are not.

Reply:
My reasoning was that `Bytes.create` is an allocation point where `b.position` may change, so I tried to consistently use `old_pos` before and `b.position` after. I could keep using `b.position` everywhere except in the final `assert`, but I felt that doing this was more in line with the treatment of `(old_)len`. What would you suggest?

Reply:
Ah, I didn't get that this was reasoning about whether code allocates. There also seems to be an assumption that the ref optimisation is working, so that line 64 is not an allocation.

I'm still not sure I understand the point here. Is the idea that unsafe mutation of the buffer will trigger assertion failures instead of segfaults? It seems possible for the assertions to be triggered by calling `Buffer.reset` inside a finaliser, is that the intent?

Reply:
> There also seems to be an assumption that the ref optimisation is working, so that line 64 is not an allocation.

The behavior is easier to reason about if there is no allocation there, but actually I think whether there is an allocation or not does not matter for correctness of the new implementation. All reads to the `b` field happen before both allocations, and after both allocations.

> Is the idea that unsafe mutation of the buffer will trigger assertion failures instead of segfaults? It seems possible for the assertions to be triggered by calling Buffer.reset inside a finaliser, is that the intent?

Yes -- to both. I think that other changes from a finaliser could be problematic (such as to the buffer and advanced `b.position` more than planned); but in any case, the assertions let us be sure that the invariants we care about for memory-safety are preserved when/if this function return.

Reply:
> The behavior is easier to reason about if there is no allocation there, but actually I think whether there is an allocation or not does not matter for correctness of the new implementation. All reads to the fields of b happen before both allocations, and after both allocations.

Ah, you're right!

> Yes -- to both. I think that other changes from a finaliser could be problematic (such as adding to the buffer and advancing b.position); but in any case, the assertions let us be sure that the invariants we care about for memory-safety are preserved when/if this function return.

OK, this makes sense. I was confused by the comments ("preserve the invariant" / "be memory-safe") - the invariants of this code can still be broken in a memory-unsafe way, but the change means that such breakage is always detected with an assertion failure. This is a clear improvement and I'm happy to merge, but I'd prefer if the comments were changed.

Reply:
Thanks for the feedback, I'll try to reword the comments.

Reply:
@stedolan I gave more detailed explanation of what is assumed and why in 30f6ef4f6e8807e851d5e02e78283859cce0b7c3 (now pushed into this PR). Does that clarify the situation for you?

Reply:
Yep, this clarifies things. Thanks!